fix issue #40 (case initializer promoted to return Optional)

mayoff opened the pull request:
Swift promotes (A) -> B to (A) -> B? if necessary to make a statement
type-check. This can lead to an enum case initializer being promoted and
making extract erroneously return nil. Fixing it is hard, so instead I'm
detecting the problem and asserting.

Fixes #40.

stephencelis commented:
@mayoff I took a quick stab at fixing the problem here: https://github.com/pointfreeco/swift-case-paths/compare/fix-optional-promotion

I'm sure it could be cleaned up (and you may have suggestions on how to do so), but do you see anything wrong with this approach?

mayoff commented:
@stephencelis Unfortunately, your approach won't work. I've added a failing test case for it: https://github.com/mayoff/swift-case-paths/commit/7cb4382c5c06b9edac31350a93e252220bf3d0fd

I added the test case to my fix and it passes.

My fix has some code duplication because I needed to overload `extract` to catch the `Optional` case and, in the inner closure, unwrap optionals in two places. I didn't see a pleasant way to extract the common bits.

mayoff commented:
To clarify why your solution doesn't work:

In my test case, both `E` (the enum type) and `E?` are stored in a single word:

- `E?.none` is stored as zero. Zero is not a valid `E`.
- `E?.some(.c1(o))` is stored as the pointer to `o`.
- `E?.some(.c2(o))` is stored as the pointer to `o`, with its high bit set.

Your fix uses the `E?` value witness table to project an `E?` to an `E` and then blindly treats the `E` as a `TestObject` pointer. Since you don't use the `E` value witness table to project the `E` to a `TestObject`, you don't strip the tag bit from the `E.c2` case and get an invalid `TestObject` pointer. (This need to double-project is why I initially decided this was too hard to bother fixing!)

stephencelis commented on Sources/CasePaths/EnumReflection.swift:
```diff
@@ -75,6 +86,16 @@ public func extract<Root, Value>(_ embed: @escaping (Value) -> Root) -> (Root) -
+    wrappedType != Value.self,
+    EnumMetadata(wrappedType) != nil
+  {
+    assertionFailure("embed parameter must not be promoted to return Optional (https://github.com/pointfreeco/swift-case-paths/issues/40)")
```
Is it still possible to hit this code path?

mayoff replied:
Seems unlikely to be hit without heroic effort.

stephencelis replied:
I wonder if we should maybe not worry about maintaining it and the associated runtime code, then. What do you think?

mayoff replied:
I think I could cook up a scenario involving generics that would call the wrong overload... but I really don't want to. That's what I meant by heroic effort. 😅 I'd personally leave the assertion there, just in case.

stephencelis commented:
> Your fix uses the `E?` value witness table to project an `E?` to an `E` (which is actually a no-op) and then blindly treats the `E` as a `TestObject` pointer. Since you don't use the `E` value witness table to project the `E` to a `TestObject`, you don't strip the tag bit from the `E.c2` case and get an invalid `TestObject` pointer. (This need to double-project is why I initially decided this was too hard to bother fixing!)

I was afraid of that. Thanks for prepping the failing case!

stephencelis commented on Sources/CasePaths/EnumReflection.swift:
```diff
@@ -11,6 +11,13 @@ extension CasePath {
     )
   }
+
+  public static func optionalCase<Wrapped>(_ embed: @escaping (Value) -> Wrapped?) -> CasePath where Root == Wrapped? {
```
Can this be an overload, as well?
```suggestion
  public static func `case`<Wrapped>(_ embed: @escaping (Value) -> Wrapped?) -> CasePath where Root == Wrapped? {
```

mayoff replied:
I tried that and it didn't work.

stephencelis replied:
Could be we need to specify generics differently. If you can remember a case that didn't work I can mess with it.

mayoff replied:
Rename it to `case` and change the `/` overload that calls it. In Xcode 13 beta 3, the `/` operator then calls the wrong `case` overload.

stephencelis commented on Sources/CasePaths/Operators.swift:
```diff
@@ -26,6 +26,12 @@ public prefix func / <Root, Value>(
+public prefix func / <Root, Value>(
+  embed: @escaping (Value) -> Root?
+) -> CasePath<Root?, Value> {
+  .optionalCase(embed)
```
(Would have to rename this:)
```suggestion
  .case(embed)
```

mayoff replied:
If it works for you, ok. I seem to recall when I tried overloading the name, it called the non-optional version somewhere that it should have called the optional version.

stephencelis replied:
Just wondering how that would be possible since the module-level `extract` function and `/` operators seem to overload just fine. Is there something strange with the generic context? Can you cook up a failure that I can take a look at?

mayoff replied:
I just double-checked. If you rename `optionalCase` to `case`, the `/` overload that currently calls `optionalCase` calls the wrong overload of `case`.

mayoff replied:
... and then it hits the “embed parameter must not be promoted” assertion in `extract`.

stephencelis replied:
Dang, yeah. For whatever reason it seems that optional promotion does its thing before overload resolution?

stephencelis commented:
Brandon and I discussed and think we should support this, even though overloads can cause the compiler to churn in some cases.

One other place that may need an overload is where we treat case path syntax as extract functions, e.g.:

```swift
results.compactMap(/Result.success)
```

See: https://github.com/pointfreeco/swift-case-paths/blob/173007350b9f5202fa7d9c2842407a363ca076a2/Sources/CasePaths/Operators.swift#L84-L89

It might be an unlikely code path, but if `results: [Result<Int, Never>?]`, then I think the above would fail at runtime? 

Hopefully this is the extent of overloads for now, and maybe we'll have a more focussed API in time for a 1.0.

stephencelis commented:
@mayoff Thanks again for exploring this! We're going to merge as is and I'll make a few changes after, mainly I'm going to make `optionalCase` internal/private, and see if I can address the case path expression overload, as well.

I think now that CasePaths has matured we will probably eliminate a lot of the API surface area. The static `.case`s seem unnecessary. We introduced them because we though the prefix operator might scare some folks off, but we haven't seen anyone prefer them to `/` in the wild. And at that point, the top-level `extract` functions seem unnecessary when it's easy enough to use a case path expression or case path. We'll look into deprecating these soon, but if you (or anyone) has any concerns with that direction, let us know!
